pkg/process/fork.go

```diff
@@ -0,0 +1,264 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package process
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"io"
+	"os"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/pkg/errors"
+)
+
+const (
+	errKill           = "failed to kill process"
+	errNotInitialized = "process not initialized"
+	errNotBuffered    = "stream not buffered"
+)
+
+func New(pathExec string, args []string, cwd string, autoStart, bufferStreams, logStreams bool,
+	logger logging.Logger) (*Info, error) {
+	result := &Info{
+		cmd:        exec.Command(pathExec, args...),
+		wgStreams:  &sync.WaitGroup{},
+		chStopped:  make(chan bool),
+		logger:     logger,
+		logStreams: logStreams,
+	}
+	if cwd != "" {
+		result.cmd.Dir = cwd
+	}
+	if bufferStreams {
+		result.buffStdout = &bytes.Buffer{}
+		result.buffStderr = &bytes.Buffer{}
+	}
+
+	stdout, err := result.cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+	result.stdout = stdout
+
+	stderr, err := result.cmd.StderrPipe()
+	if err != nil {
+		return nil, err
+	}
+	result.stderr = stderr
+
+	if autoStart {
+		return result, result.cmd.Start()
+	}
+	return result, nil
+}
+
+type lineProcessor func(line string, logger logging.Logger)
+
+type Info struct {
+	cmd        *exec.Cmd
+	wgStreams  *sync.WaitGroup
+	chStopped  chan bool
+	stdout     io.ReadCloser
+	stderr     io.ReadCloser
+	logger     logging.Logger
+	buffStdout *bytes.Buffer
+	buffStderr *bytes.Buffer
+	logStreams bool
+}
+
+func (pi *Info) StartStdout(processor lineProcessor) {
+	pi.processStream(pi.stdout, processor)
+}
+
+func (pi *Info) LogStdout() {
+	pi.StartStdout(nil)
+}
+
+func (pi *Info) StartStderr(processor lineProcessor) {
+	pi.processStream(pi.stderr, processor)
+}
+
+func (pi *Info) LogStderr() {
+	pi.StartStderr(nil)
+}
+
+func (pi *Info) Log() {
+	pi.LogStdout()
+	pi.LogStderr()
+}
+
+func (pi *Info) StdoutAsString() (string, error) {
+	if pi.buffStdout == nil {
+		return "", errors.New(errNotBuffered)
+	}
+	return pi.buffStdout.String(), nil
+}
+
+func (pi *Info) StderrAsString() (string, error) {
+	if pi.buffStderr == nil {
+		return "", errors.New(errNotBuffered)
+	}
+	return pi.buffStderr.String(), nil
+}
+
+func (pi *Info) Run(ctx context.Context) error {
+	if pi.cmd == nil {
+		return errors.New(errNotInitialized)
+	}
+	// when this routine is scheduled, if ctx is already done, do not start the process
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+
+	default:
+	}
+
+	chErr := make(chan error, 1)
+	defer close(chErr)
+
+	go func() {
+		if err := pi.cmd.Start(); err != nil {
+			chErr <- err
+		}
+		pi.LogStdout()
+		pi.LogStderr()
+		chErr <- pi.WaitError()
+	}()
+
+	select {
+	case <-ctx.Done():
```

Comment:
Interesting point. When we are considering cleanup logic in a separate PR, we may consider sending a `SIGINT`.

Reply:
A way to kill the ongoing process would be nice for deletion, too. Right now, if deletion call comes in and there is an ongoing `apply` operation, we just wait for it to complete instead of sending a kill signal to `apply` process.